File: server/src/psychosims_server/logger.py
"""Cross-stack logger for the Python server.

Emits the canonical log-line schema defined in ``docs/code/LOGGING.md``.
Mode (text vs JSON) and minimum level are read from the config authority at
import time; in tests they default to text/info.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Mapping


class Level(str, Enum):
    DEBUG = "debug"
    INFO = "info"
    SUCCESS = "success"
    WARN = "warn"
    ERROR = "error"
    FATAL = "fatal"


_LEVEL_EMOJI = {
    Level.DEBUG: "🔍",
    Level.INFO: "ℹ️",
    Level.SUCCESS: "✅",
    Level.WARN: "⚠️",
    Level.ERROR: "❌",
    Level.FATAL: "🛑",
}


class ErrorKind(str, Enum):
    USER = "user"
    SYSTEM = "system"
    EXTERNAL = "external"
    OFFLINE = "offline"


class PsyLogger:
    """Structured logger that matches the Dart/C++ logger output exactly."""

    def __init__(
        self,
        min_level: Level = Level.INFO,
        json_mode: bool = False,
        correlation_id: str = "none",
    ) -> None:
        self.min_level = min_level
        self.json_mode = json_mode
        self.correlation_id = correlation_id

    def _emit(self, line: str) -> None:
        sys.stderr.write(line + "\n")

    def _level_index(self, level: Level) -> int:
        return list(Level).index(level)

    def log(
        self,
        level: Level,
        scope: str,
        event: str,
        kv: Mapping[str, Any] | None = None,
        error_kind: ErrorKind | None = None,
        message: str | None = None,
    ) -> None:
        if self._level_index(level) < self._level_index(self.min_level):
            return

        ts = datetime.now(timezone.utc).isoformat()
        safe_scope = scope.lower()
        safe_event = event.lower()

        if self.json_mode:
            payload: dict[str, Any] = {
                "ts": ts,
                "level": level.value,
                "scope": safe_scope,
                "event": safe_event,
                "correlation_id": self.correlation_id,
            }
            if kv:
                payload["kv"] = dict(kv)
            if error_kind:
                payload["error_kind"] = error_kind.value
            self._emit(json.dumps(payload, separators=(",", ":"), sort_keys=True))
        else:
            parts = [
                f"{_LEVEL_EMOJI[level]} {level.value.upper():7} "
                f"[{ts}] [{safe_scope}] {safe_event:24}",
                f"corr={self.correlation_id}",
            ]
            if kv:
                parts.append("| " + " ".join(f"{k}={v}" for k, v in kv.items()))
            if error_kind:
                parts.append(f"| error_kind={error_kind.value}")
            if message:
                parts.append(f"| {message}")
            self._emit(" ".join(parts))

    def debug(self, scope: str, event: str, **kwargs: Any) -> None:
        self.log(Level.DEBUG, scope, event, **kwargs)

    def info(self, scope: str, event: str, **kwargs: Any) -> None:
        self.log(Level.INFO, scope, event, **kwargs)

    def success(self, scope: str, event: str, **kwargs: Any) -> None:
        self.log(Level.SUCCESS, scope, event, **kwargs)

    def warn(self, scope: str, event: str, **kwargs: Any) -> None:
        self.log(Level.WARN, scope, event, **kwargs)

    def error(
        self, scope: str, event: str, error_kind: ErrorKind | None = None, **kwargs: Any
    ) -> None:
        self.log(Level.ERROR, scope, event, error_kind=error_kind or ErrorKind.SYSTEM, **kwargs)

    def fatal(
        self, scope: str, event: str, error_kind: ErrorKind | None = None, **kwargs: Any
    ) -> None:
        self.log(Level.FATAL, scope, event, error_kind=error_kind or ErrorKind.SYSTEM, **kwargs)


def default_logger() -> PsyLogger:
    """Build a logger from environment; used until the config authority is wired."""
    min_level = Level(os.environ.get("PSY_LOG_LEVEL", "info"))
    json_mode = os.environ.get("PSY_LOG_MODE", "text").lower() == "json"
    correlation_id = os.environ.get("PSY_CORRELATION_ID", "none")
    return PsyLogger(min_level=min_level, json_mode=json_mode, correlation_id=correlation_id)


# Module-level default instance for convenience. Prefer injecting a configured
# instance once the config authority is available.
logger = default_logger()

# Do not expose a plain logging.Logger that could bypass the schema.
__all__ = ["PsyLogger", "Level", "ErrorKind", "default_logger", "logger"]

File: app/lib/shared/logger.dart
import 'dart:convert';
import 'dart:developer' as developer;

/// Cross-stack logger emitting the canonical log-line schema defined in
/// `docs/code/LOGGING.md`.
///
/// The logger is intentionally small and dependency-free. The rendering mode
/// (text vs JSON) and minimum level are supplied by the config authority;
/// until config is available it defaults to dev text and [LogLevel.info].
enum LogLevel {
  debug('debug', '🔍'),
  info('info', 'ℹ️'),
  success('success', '✅'),
  warn('warn', '⚠️'),
  error('error', '❌'),
  fatal('fatal', '🛑');

  const LogLevel(this.name, this.emoji);
  final String name;
  final String emoji;

  bool operator >=(LogLevel other) => index >= other.index;
}

enum ErrorKind { user, system, external, offline }

class PsyLog {
  PsyLog({
    this.minLevel = LogLevel.info,
    this.jsonMode = false,
    this.correlationId = 'none',
  });

  final LogLevel minLevel;
  final bool jsonMode;
  final String correlationId;

  void log(
    LogLevel level,
    String scope,
    String event, {
    Map<String, Object?> kv = const {},
    ErrorKind? errorKind,
    String? message,
  }) {
    if (level.index < minLevel.index) return;

    final timestamp = DateTime.now().toUtc().toIso8601String();
    final safeScope = scope.toLowerCase();
    final safeEvent = event.toLowerCase();

    if (jsonMode) {
      final payload = <String, Object?>{
        'ts': timestamp,
        'level': level.name,
        'scope': safeScope,
        'event': safeEvent,
        'correlation_id': correlationId,
        if (kv.isNotEmpty) 'kv': kv,
        if (errorKind != null) 'error_kind': errorKind.name,
      };
      _emit(jsonEncode(payload));
    } else {
      final sb = StringBuffer()
        ..write('${level.emoji} ${level.name.toUpperCase().padRight(7)} ')
        ..write('[$timestamp] ')
        ..write('[$safeScope] ')
        ..write(safeEvent.padRight(24))
        ..write(' corr=$correlationId');
      if (kv.isNotEmpty) {
        sb.write(
            ' | ${kv.entries.map((e) => '${e.key}=${e.value}').join(' ')}');
      }
      if (errorKind != null) sb.write(' | error_kind=${errorKind.name}');
      if (message != null) sb.write(' | $message');
      _emit(sb.toString());
    }
  }

  void debug(String scope, String event,
          {Map<String, Object?> kv = const {}, String? message}) =>
      log(LogLevel.debug, scope, event, kv: kv, message: message);
  void info(String scope, String event,
          {Map<String, Object?> kv = const {}, String? message}) =>
      log(LogLevel.info, scope, event, kv: kv, message: message);
  void success(String scope, String event,
          {Map<String, Object?> kv = const {}, String? message}) =>
      log(LogLevel.success, scope, event, kv: kv, message: message);
  void warn(String scope, String event,
          {Map<String, Object?> kv = const {}, String? message}) =>
      log(LogLevel.warn, scope, event, kv: kv, message: message);
  void error(String scope, String event,
          {Map<String, Object?> kv = const {},
          ErrorKind? errorKind,
          String? message}) =>
      log(LogLevel.error, scope, event,
          kv: kv, errorKind: errorKind ?? ErrorKind.system, message: message);
  void fatal(String scope, String event,
          {Map<String, Object?> kv = const {},
          ErrorKind? errorKind,
          String? message}) =>
      log(LogLevel.fatal, scope, event,
          kv: kv, errorKind: errorKind ?? ErrorKind.system, message: message);

  void _emit(String line) {
    // Use developer.log in debug builds to keep output inspectable;
    // in release/JSON mode we still avoid raw `print`.
    developer.log(line);
  }
}
